src/gsSolver/gsParameterDomainPreconditioners.hpp

```diff
@@ -0,0 +1,438 @@
+/** @file gsParameterDomainPreconditioners.hpp
+
+    @brief Provides preconditioners that live on the parameter domain.
+
+    This file is part of the G+Smo library.
+
+    This Source Code Form is subject to the terms of the Mozilla Public
+    License, v. 2.0. If a copy of the MPL was not distributed with this
+    file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+    Author(s): S. Takacs, C. Hofreither
+*/
+#pragma once
+
+#include <gsSolver/gsParameterDomainPreconditioners.h>
+#include <gsSolver/gsSumOfOperatorsOp.h>
+#include <gsSolver/gsProductOfOperatorsOp.h>
+#include <gsSolver/gsKroneckerOp.h>
+#include <gsSolver/gsMatrixOp.h>
+#include <gsMatrix/gsKronecker.h>
+#include <gsTensor/gsTensorDomainIterator.h>
+#include <gsTensor/gsTensorDomainBoundaryIterator.h>
+#include <gsTensor/gsTensorBasis.h>
+
+namespace gismo
+{
+
+template<typename T>
+void gsParameterDomainPreconditioners<T>::init()
+{
+
+    GISMO_ASSERT( m_dirichlet == dirichlet::elimination || m_dirichlet == dirichlet::none,
+                  "Only the dirichlet strategies dirichlet::elimination and dirichlet::none are currently implemented." );
+    // We do not provide:
+    //   dirichlet::penalize
+    //   dirichlet::nitsche
+    //   dirichlet::eliminatNormal
+
+}
+
+namespace {
+
+template <typename T>
+int estimateNonzerosPerRow(const gsBasis<T>& basis)
+{
+    int nnz = 1;
+    for (int i = 0; i < basis.dim(); ++i)
+        nnz *= 2 * basis.degree(i) + 1;
+    return nnz;
+}
+
+template <typename T>
+void localToGlobal(const gsMatrix<T>& localMat, const gsMatrix<unsigned>& localDofs, gsSparseMatrix<T>& globalMat)
+{
+    const int numActive = localDofs.rows();
+
+    for (index_t i = 0; i < numActive; ++i)
+    {
+        const int ii = localDofs(i,0);
+        for (index_t j = 0; j < numActive; ++j)
+        {
+            const int jj = localDofs(j,0);
+            globalMat.coeffRef(ii, jj) += localMat(i, j);
+        }
+    }
+}
+
+} // anonymous namespace
+
+template<typename T>
+gsSparseMatrix<T> gsParameterDomainPreconditioners<T>::assembleMass(const gsBasis<T>& basis)
+{
+    const int n = basis.size();
+
+    gsSparseMatrix<T> M(n, n);
+    M.reserve( gsVector<int>::Constant(n, estimateNonzerosPerRow(basis)) );
+
+    gsMatrix<T> localMass;
+
+    // Make domain element iterator
+    typename gsBasis<T>::domainIter domIt = basis.makeDomainIterator();
+
+    // Set the number of integration points for each element
+    gsVector<index_t> numQuadNodes( basis.dim() );
+    for (int i = 0; i < basis.dim(); ++i)
+        numQuadNodes[i] = basis.degree(i) + 1;
+    domIt->computeQuadratureRule( numQuadNodes );
```

Comment:
domIt->computeQuadratureRule( numQuadNodes ); 
this function is deprecated and should not be used..

Reply:
i cannot find related information.